### ISSUE TITLE
InternalIPKernel: a second create and destroy fails.

### ISSUE THREAD

**Issue by mdickinson** [enthought/envisage#187]
The following code fails for me (Python 3.6 / macOS):

```python
from envisage.plugins.ipython_kernel.internal_ipkernel import InternalIPKernel

def create_and_destroy_kernel():
    kernel = InternalIPKernel()
    kernel.init_ipkernel(gui_backend=None)
    kernel.shutdown()

create_and_destroy_kernel()
create_and_destroy_kernel()
```

with the following exception:
```
(envisage-test-3.6-null)bash-3.2$ python create_and_destroy.py 

To connect another client to this kernel, use:
    --existing kernel-41771.json
Traceback (most recent call last):
  File "create_and_destroy.py", line 9, in <module>
    create_and_destroy_kernel()
  File "create_and_destroy.py", line 5, in create_and_destroy_kernel
    kernel.init_ipkernel(gui_backend=None)
  File "/Users/mdickinson/Enthought/ETS/envisage/envisage/plugins/ipython_kernel/internal_ipkernel.py", line 91, in init_ipkernel
    self.ipkernel = gui_kernel(gui_backend)
  File "/Users/mdickinson/Enthought/ETS/envisage/envisage/plugins/ipython_kernel/internal_ipkernel.py", line 35, in gui_kernel
    kernel.initialize(argv)
  File "<decorator-gen-121>", line 2, in initialize
  File "/Users/mdickinson/.edm/envs/envisage-test-3.6-null/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/Users/mdickinson/.edm/envs/envisage-test-3.6-null/lib/python3.6/site-packages/ipykernel/kernelapp.py", line 469, in initialize
    self.init_sockets()
  File "/Users/mdickinson/.edm/envs/envisage-test-3.6-null/lib/python3.6/site-packages/ipykernel/kernelapp.py", line 239, in init_sockets
    self.shell_port = self._bind_socket(self.shell_socket, self.shell_port)
  File "/Users/mdickinson/.edm/envs/envisage-test-3.6-null/lib/python3.6/site-packages/ipykernel/kernelapp.py", line 181, in _bind_socket
    s.bind("tcp://%s:%i" % (self.ip, port))
  File "zmq/backend/cython/socket.pyx", line 550, in zmq.backend.cython.socket.Socket.bind
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Address already in use
```

**Comment by mdickinson** [comment]
Likely related to #186: if we do proper cleanup here, we'll be leaking fewer file descriptors.

**Comment by mdickinson** [comment]
The initial cause looks straightforward enough: the `IPKernelApp` class (`ipykernel.kernelapp.IPKernelApp`) has an `init_sockets` method that sets up various ZMQ sockets, but at no point are those sockets destroyed. It's conceivable that the best solution may be either to write our own `IPKernelApp` class that includes proper cleanup, or to try to contribute to the upstream `ipykernel`.

There's a catch with the latter, though: we're using `ipykernel 4.10.0`, which is fairly old; there's a bugfix release `ipykernel 4.10.1` upstream, and an `ipykernel 5.x` series that would need major work to use (it takes an entirely different approach to the event loop integration).

**Comment by mdickinson** [comment]
Closed by #188.
